core/kernel/tee_ta_manager.c

```diff
@@ -823,6 +823,161 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
+				       uint32_t cancel_req_to,
+				       struct tee_ta_param *param)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_ctx *ctx = NULL;
+	struct ts_ctx *ts_ctx = NULL;
+
+	ts_ctx = s->ts_sess.ctx;
+	if (ts_ctx)
+		ctx = ts_to_ta_ctx(ts_ctx);
+
+	if (ctx->panicked)
+		return TEE_ERROR_TARGET_DEAD;
+
+	if (tee_ta_try_set_busy(ctx)) {
+		s->param = param;
+		set_invoke_timeout(s, cancel_req_to);
+		res = ts_ctx->ops->dump_mem_stats(&s->ts_sess);
+		s->param = NULL;
+		tee_ta_clear_busy(ctx);
+	} else {
+		/* Deadlock avoided */
+		res = TEE_ERROR_BUSY;
+	}
+
+	return res;
+}
+
+TEE_Result tee_ta_instance_stats(void *buff,
+				 uint32_t *buff_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_param param = { };
+	struct tee_ta_dump_stats *dump_stats = NULL;
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	uint32_t cnt = 0;
+	uint32_t size = 0;
+	uint32_t actual_size = 0;
+	uint32_t stats_num = 0;
+
+	if (!buff_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&tee_ta_mutex);
+
+	/*
+	 * Go through all available TA and calc out the actual buffer size.
+	 */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link)
+		actual_size += sizeof(struct tee_ta_dump_stats);
+
+	if (!buff || *buff_size == 0) {
+		/*
+		 * Buff is null or pass size with 0 means caller want to get
+		 * the size of buff.
+		 */
+		*buff_size = actual_size;
+		mutex_unlock(&tee_ta_mutex);
+		return res;
+	}
+
+	/*
+	 * If the buffer size is smaller than the actual size,
+	 * update *buff_size and return TEE_ERROR_SHORT_BUFFER
+	 */
+	size = *buff_size;
+	if (size < actual_size) {
+		*buff_size = actual_size;
+		res = TEE_ERROR_SHORT_BUFFER;
+		mutex_unlock(&tee_ta_mutex);
+		return res;
+	}
+
+	dump_stats = (struct tee_ta_dump_stats *)buff;
+	nsec_sessions_list_head(&open_sessions);
+	/*
+	 * Scan all sessions opened from secure side by searching through
+	 * all available TA instances and for each context, scan all opened
+	 * sessions.
+	 */
+	/* Context of a loaded TA */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link) {
+		memcpy(&dump_stats[stats_num].uuid, &ctx->ts_ctx.uuid,
+		       sizeof(ctx->ts_ctx.uuid));
+		dump_stats[stats_num].panicked = ctx->panicked;
+		dump_stats[stats_num].is_user_ta =
+			is_user_ta_ctx(&ctx->ts_ctx);
+		cnt = 0;
+		TAILQ_FOREACH(sess, open_sessions, link) {
+			if (sess->ts_sess.ctx == &ctx->ts_ctx) {
+				if (cnt == MAX_DUMP_SESS_NUM)
+					break;
+
+				dump_stats[stats_num].sess_id[cnt] =
+					sess->id;
+				cnt++;
+			}
+		}
+
+		dump_stats[stats_num].sess_num = cnt;
+		stats_num++;
+	}
+	mutex_unlock(&tee_ta_mutex);
+
+	/*
+	 * Go through the dump context to get heap stats if it's user.
+	 */
+	for (uint32_t i = 0; i < stats_num; i++) {
+		struct tee_ta_dump_stats *stats = &dump_stats[i];
+
+		if (!stats->is_user_ta || stats->panicked)
+			continue;
+
+		/*find a session from TA dump context*/
+		sess = NULL;
+		for (uint32_t j = 0; j < stats->sess_num; j++) {
```

Comment:
You're reading from non-secure shared memory here. What if that has been changed since we last wrote it?

Reply:
make sense...shall not save the intermediate context in dump buffer.